File: app/src/at-trivo/java/asiantech/internship/summer/recycler_view/timeline_recycler_view/TimelineAdapter.java
package asiantech.internship.summer.recycler_view.timeline_recycler_view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import asiantech.internship.summer.R;
import asiantech.internship.summer.recycler_view.model.Timeline;

public class TimelineAdapter extends RecyclerView.Adapter {

    private final List<Timeline> mDataset;
    private Context mContext;
    private TimelineViewHolder.TimeLineViewHolderListener mListener;
    private final int VIEW_TYPE_ITEM = 1;
    private final int VIEW_TYPE_PROGRESS_BAR = 0;
    private final int VIEW_TYPE_HEADER = 2;

    public TimelineAdapter(List<Timeline> dataset, Context context, TimelineViewHolder.TimeLineViewHolderListener listener) {
        mContext = context;
        mDataset = dataset;
        mListener = listener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        switch (viewType) {
            case VIEW_TYPE_ITEM:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_holder_timeline, parent, false);
                return new TimelineViewHolder(view, mListener);
            case VIEW_TYPE_PROGRESS_BAR:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_holder_progress_bar, parent, false);
                return new ProgressBarViewHolder(view);
            case VIEW_TYPE_HEADER:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_holder_header, parent, false);
                return new HeaderViewHolder(view);
            default:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_holder_timeline, parent, false);
        }
        return new TimelineViewHolder(view, mListener);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int type = getItemViewType(position);
        switch (type) {
            case VIEW_TYPE_ITEM:
                onBindViewHolder(((TimelineViewHolder) holder), position);
                break;
            case VIEW_TYPE_PROGRESS_BAR:
                ((ProgressBarViewHolder) holder).getProgressBar().setIndeterminate(true);
                break;
        }
    }

    private void onBindViewHolder(TimelineViewHolder timelineViewHolder, int position) {
        timelineViewHolder.setPosition(position);
        Timeline timeline = mDataset.get(position);
        timelineViewHolder.getImgAuthor().setImageResource(timeline.getAuthor().getProfileImageId());
        timelineViewHolder.getTvAuthorName().setText(timeline.getAuthor().getName());
        timelineViewHolder.getImgTimeline().setImageResource(timeline.getTimelineImageId());
        StringBuilder stringBuilder = new StringBuilder().append(timeline.getLoveNumber()).append(mContext.getResources().getString(R.string.like));
        timelineViewHolder.getTvLoveNumber().setText(stringBuilder);
        timelineViewHolder.getTvDescription().setText(timeline.getDescription());
        boolean isLoved = timeline.isLoved();
        if (isLoved) {
            timelineViewHolder.getImgHeart().setImageResource(R.drawable.ic_heart_filled);
        } else {
            timelineViewHolder.getImgHeart().setImageResource(R.drawable.ic_heart);
        }
    }

    @Override
    public int getItemCount() {
        return mDataset.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (mDataset.get(position) == null) {
            if (position == 0) {
                return VIEW_TYPE_HEADER;
            }
            return VIEW_TYPE_PROGRESS_BAR;
        }
        return VIEW_TYPE_ITEM;
    }
}

File: app/src/at-trivo/java/asiantech/internship/summer/recyclerviewkotlin/timelinerecyclerview/TimelineAdapter.kt
package asiantech.internship.summer.recyclerviewkotlin.timelinerecyclerview

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import asiantech.internship.summer.R
import asiantech.internship.summer.recyclerviewkotlin.TimelineViewHolderInteractListener
import asiantech.internship.summer.recyclerviewkotlin.model.Timeline

const val VIEW_TYPE_TIMELINE = 1
const val VIEW_TYPE_HEADER = 2
const val VIEW_TYPE_PROGRESS_BAR = 3

class TimelineAdapter(private val dataset: MutableList<Timeline?>, private val listener: TimelineViewHolderInteractListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view: View
        when (viewType) {
            VIEW_TYPE_TIMELINE -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_timeline, parent, false)
                return TimelineViewHolder(view, listener)
            }
            VIEW_TYPE_PROGRESS_BAR -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_progress_bar, parent, false)
                return ProgressBarViewHolder(view)
            }
            VIEW_TYPE_HEADER -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_header, parent, false)
                return HeaderViewHolder(view)
            }
            else -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_timeline, parent, false)
                return TimelineViewHolder(view, listener)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val timeline: Timeline? = dataset[position]
        if (holder is TimelineViewHolder && timeline is Timeline) {
            onBindViewHolder(holder, timeline, position)
        } else {
            if (holder is ProgressBarViewHolder) {
                holder.progressBar.isIndeterminate = true
            }
        }
    }

    private fun onBindViewHolder(timelineViewHolder: TimelineViewHolder, timeline: Timeline, position: Int) {
        timelineViewHolder.imgAuthor.setImageResource(timeline.author.profileImageId)
        timelineViewHolder.imgTimeline.setImageResource(timeline.imageId)
        timelineViewHolder.tvAuthorName.text = timeline.author.name
        timelineViewHolder.tvDescription.text = timeline.description
        timelineViewHolder.tvLoveNumber.text = timeline.loveNumber.toString()
        timelineViewHolder.dataPosition = position
        val heartImageId = if (timeline.isLoved) R.drawable.ic_heart_filled else R.drawable.ic_heart
        timelineViewHolder.imgHeart.setImageResource(heartImageId)
    }


    override fun getItemViewType(position: Int): Int {
        dataset[position]?.let {
            return VIEW_TYPE_TIMELINE
        }
        return if (position == 0) VIEW_TYPE_HEADER else VIEW_TYPE_PROGRESS_BAR
    }

    override fun getItemCount(): Int = dataset.size
}